---
title: "Fig-1a"
author: "Andrea Mariossi"
format: html
editor: visual
editor_options: 
  chunk_output_type: inline
---

### setup

```{r}
rm(list = ls())

# Package loading
suppressPackageStartupMessages({
  # Data manipulation & file import
  library(tidyverse)
  library(janitor)
  library(readxl)
  # Statistical analysis
  library(rstatix)
  library(Rmisc)
  # Visualization
  library(ggpubr)
  library(ggbeeswarm)
  library(ggdist)
  library(gghalves)
  library(patchwork)
  library(paletteer)
  #rna-seq
  library(SummarizedExperiment)
  library(readxl)
  library(Cairo)
})

```

### configuration

```{r}

# ===== path =====
file_path <- "/Users/andrea/Princeton Dropbox/Andrea Mariossi/4.paper_archive/2025_thyroid_pathway_&_metamorphosis/3.1.ScienceAdvances_submission_rev/tables/"

# ===== theme & style =====
custom_theme <- function() {
  theme(
    axis.ticks = element_line(size = 0.5),
    axis.ticks.length = unit(0.09, "cm"),
    legend.key.height = unit(20, "pt"),
    axis.line = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA, size = 1),
    panel.background = element_blank(),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.background = element_blank(),
    legend.box.background = element_blank(),
    legend.key = element_blank(),
    strip.background = element_blank(),
  )
}

# ===== set theme =====
theme_set(theme_pubr(base_size = 13,
                     base_family = "Helvetica") + custom_theme())

# ===== extra bits =====
no_legend <- theme(legend.position = "none")
axis_text_45 <- theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

```

```{r}

# ===== developmental stages =====
stage_level_hu <- c("fertE","cell2","cell8","cell16","cell32","cell64","iniG", "midG", "earN", "latN", "earTI", "midTII", "latTII", "earLarva","latLarva", "earRot","latRot","earJv", "midJv","latJv","adult")

# ===== developmental classification =====
stage_category_hu <- ifelse(stage_level_hu %in% c("fertE","cell2","cell8","cell16","cell32","cell64"), "early_dev",
                  ifelse(stage_level_hu %in% c("iniG","midG", "earN", "latN", "earTI", "midTII"), "med_dev",
                  ifelse(stage_level_hu %in% c("latTII", "earLarva","latLarva"), "late_dev",
                         "meta")))

# Stage classification
stage_lookup_hu <- 
  tibble(stage_level_hu, stage_category_hu) |> 
  mutate(stage_level_hu = as_factor(stage_level_hu),
         stage_category_hu = as_factor(stage_category_hu))

levels(stage_lookup_hu$stage_category_hu)

# ===== time mapping =====

# -------------------------------------
# To map stage to time hours post fertilization
# -------------------------------------

time_map.df <- tibble(
  stage = c("st.1", "st.2", "st.4", "st.5", "st.6", "st.8", "st.10", "st.12", "st.14", "st.16",
            "st.19", "st.22", "st.24","st.26", "st.27", "st.29", "st.35", "st.37", "st.38", "st.40", "st.45", "st.50"),
  description = c("One.cell", "2-cell", "8-cell", "late.16-cell", "late.32.cell", "64-cell", "110-cell",
                  "mid.gastrula", "early.neurula", "late.neurula", "early.tailbud.I", "mid.tailbud.II",
                  "late.tailbud.II", "hatching.larva", "early.swimming.larva", "late.swimming.larva", "early.rotation",
                  "late.rotation", "early.juvenile.I", "mid.juvenile", "late.juvenile", "adult"),
  hours = c(0.24, 0.55, 1.54, 2.39, 3.12, 4, 4.33, 5.39, 6.21, 7.24,
            9.19, 10.5, 13.3, 17, 19, 23, 39, 69, 96, 168, 190, 300)
)

# -------------------------------------
# to extract gene annotation 
# -------------------------------------

# ===== clean names =====
get_gene_annotation <- function(se_object) {
  return(rowData(se_object))
}

# ===== clean names =====
extract_sample_info_hu <- function(df, sample_col) {
  df %>%
    mutate(
      stage = str_extract(.data[[sample_col]], "^[^_]+"),  # Extract first part before first "_"
      stage_st = map_chr(str_split(.data[[sample_col]], "_"), ~ .x[2]),  # Extract middle part
      rep = str_extract(.data[[sample_col]], "[^_]+$")  # Extract last part after last "_"
    )
}

# ===== counts/TPM long tidy format =====
process_long_tidy <- function(counts_data, extract_sample_info_func,counts_type) {
  counts_data %>%
    rownames_to_column("genes") %>% 
    pivot_longer(-genes, names_to = "sample_id", values_to = counts_type) %>% 
    extract_sample_info_func(., "sample_id")  |> 
  left_join(stage_lookup_hu, by = c("stage" = "stage_level_hu")) |> 
  mutate(stage = factor(stage, levels = stage_level_hu),
         condititon=sample_id,
         sample_id = fct_reorder(sample_id, as.numeric(stage)))
} 


```

# Fig 1: thyroid hormone toolkit

#### Time series RNA from Hu et al. 2017

```{r}

# Load and process kallisto TPM data
hu_kallisto_data <- readRDS(paste0(file_path,
                                   '/kallisto.merged.gene_counts_length_scaled.rds'))

# ===== Inspect data structure =====
assayNames(hu_kallisto_data)
class(assay(hu_kallisto_data, "counts"))
str(assay(hu_kallisto_data, "counts"))

# ===== Extract TPM  =====
hu_tpm_kallisto <- assay(hu_kallisto_data, "abundance")
hu_tpm_kallisto |> colSums() # this is TPM

# ===== save TPM for supp table =====

# hu_tpm_kallisto |> 
#   as.data.frame() %>% 
#   rownames_to_column("gene_id") |> 
#   write_tsv("hu_tpm_kallisto_with_metadata.tsv")

# ===== long tidy format =====
hu_TPM_tidy <- process_long_tidy(hu_tpm_kallisto, extract_sample_info_hu, "TPM")

# ===== gene annotation from the SummarizedExperiment object =====
gene_annotation <- get_gene_annotation(hu_kallisto_data)

# ===== quick visualization =====
hu_TPM_tidy |> 
  filter(genes == "KY21.Chr11.1104") |> 
ggplot(aes(x = stage, color = stage, shape = rep)) +
  geom_point(aes(y = TPM), size=3, position = position_jitter(width = 0.2), alpha = 0.5) +
  no_legend +
  axis_text_45

## TH peaks before metamorphosis: we don't need the data points from latRot onward (40hpf)

```

#### Time series protein from Frese et al. 2024

```{r}

# Read Frese_2024_TMTproC
frese_prot.df <- read_excel(file.path(file_path, 
                                      "3_rna_protein_dynamics.xlsx"),
                            sheet = "TMTproC_Frese_2024_relAbuncance", 
                            cell_cols("A:L"),
                            col_types = c(rep("text", 3), rep("numeric", 9)))

# Protein values are normalize by the sum, then scale to 1
frese_prot.df <- frese_prot.df %>% 
  filter(proteinID == "KY21.Chr11.1104") %>%  # filter thyroid receptor
  pivot_longer(-c(1:4), names_to = "stage", values_to = "protein") %>%  
  group_by(proteinID) %>%
  mutate(protein_to_channel = protein * 8, # Normalize protein by TMTproC channels
         protein_normMax = protein / max(protein)  # Normalize protein by the max
         ) |> 
  pivot_longer(
    cols = c(protein, protein_to_channel, protein_normMax),
    names_to = "measurement",
    values_to = "protein"
  )

# ===== plot protein =====
ggplot(frese_prot.df, aes(x = fct_inorder(stage), y = protein, color = measurement, group = measurement)) +
  geom_line(size = 1.2) +
  geom_point(size = 2) +
  labs(x = "Developmental Stage",
       y = "Different normalization for relative expression",
       color = "Measurement"
  ) +
  facet_wrap(~measurement,scales="free_y") 

```

#### Integrate RNA & protein time series

```{r}

# ===== preprocess RNA data for integration =====
tpm_df <- 
  hu_TPM_tidy |> 
  left_join(time_map.df, by = c("stage_st" = "stage")) |> 
  # dplyr::filter(stage !="adult") |> 
  dplyr::filter(hours < 40 ) |> 
  filter(genes == "KY21.Chr11.1104") %>%
  Rmisc::summarySE(measurevar = "TPM", 
                   groupvars = c("genes", "hours","stage","stage_st")) |>
  clean_names() |> 
  group_by(genes) %>%
  mutate(
    tpm_normMax = tpm / max(tpm),
    cv = (sd / tpm) * 100, # Calculate coefficient of variation / relative standard deviation (RSD) in % | tpm is already mean(tpm)
    tpm_to1 = (tpm / sum(tpm)),  # Normalize value by the sum, then scale to 1
    # https://stats.stackexchange.com/questions/241187/calculating-standard-deviation-after-log-transformation
  )
  
sum(tpm_df$tpm_to1) # Check if the sum of tpm_to1 is 1

# ===== merge rna & prot =====
rna_protein.df <- tpm_df |>
  ungroup() |> 
    add_row(
    genes = "KY21.Chr11.1104",
    hours = 19,                    
    stage = "hatching.larva", 
    stage_st = "st.26",
    n = NA_real_,
    tpm = NA_real_,
    sd = NA_real_,
    se = NA_real_,
    ci = NA_real_,
    tpm_normMax = NA_real_,
    cv = NA_real_,
    tpm_to1 = NA_real_) |> 
  left_join(frese_prot.df %>%
              mutate(stage = dplyr::recode(stage, "larva" = "hatching.larva")) |> 
              dplyr::filter(measurement == "protein_to_channel"),
            by = c("genes" = "proteinID","stage" = "stage")) |> 
  dplyr::select(genes,hours, stage, tpm, se, protein) 
  
```

### Visualization - Panel 1A

```{r}

# calculate scaling factor for dual y-axis
# same height rna and protein for plot
scale_factor <- max(rna_protein.df$tpm, na.rm = TRUE) / max(rna_protein.df$protein, na.rm = TRUE)

thr_plot <- rna_protein.df |> 
  ggplot(aes(x = hours)) +
  
  # geom_rect(aes(xmin = 17, xmax = 23, ymin = 0, ymax = 250), fill = "grey", color = NA, alpha = 0.2) +
  
  # RNA data (primary axis)
  geom_line(data = . %>% filter(!is.na(tpm)),
            aes(y = tpm, color = "TPM"), 
            size = 1) +
  geom_point(data = . %>% filter(!is.na(tpm)),
             aes(y = tpm, color = "TPM"), 
             size = 3.5,shape = 16) +
  geom_ribbon(data = . %>% filter(!is.na(tpm)),
              aes(ymin = tpm - se, ymax = tpm + se), 
              size = 0.5, alpha = 0.5, color = NA, fill = "#00a087") +
  
  # Protein data (secondary axis)
  geom_line(data = . %>% filter(!is.na(protein)),
            aes(y = protein * scale_factor, color = "Protein"), 
            size = 1,
            linewidth = 1, linetype = "dashed") +
  geom_point(data = . %>% filter(!is.na(protein)),
             aes(y = protein * scale_factor, color = "Protein"), 
             size = 3.5,shape = 18) +
  
  # Scales and labels
  scale_x_continuous(breaks = seq(0, 40, by = 5), 
                     labels = seq(0, 40, by = 5)
                     ) +
  scale_y_continuous(name = "TPM (transcript per million)",
                     sec.axis = sec_axis(~./scale_factor, name = "Protein relative abundance")
                     ) +
  scale_color_manual(
    values = c("TPM" = "#00a087", "Protein" = "#3c5488"),
    guide = guide_legend(override.aes = list(linetype = c("dashed", "solid"),
                                             shape = c(18, 16))
                         )) +
  labs(x = "Hours post development", color = "") +
  
  # Developmental stage annotations
  annotate("segment",
           x = c(0, 17.5, 24, 27, 31), 
           xend = c(17, 23, 26, 30, 40), 
           y = 240, 
           yend = 240, 
           color = "black", size = 1.2) +
  annotate("text", 
           x = c(8.75, 20, 25, 28.5, 35), 
           y = 250, 
           label = c("Hatching", "Swi", "Adh", "Tail \nreg", "Juveniles"), 
           size = 4, color = "black") +
  
  custom_theme() +
  theme(
    legend.position = c(0.05, 0.9),
    legend.justification = c(0, 1),
    legend.text = element_text(size = 15),
    # axis.title = element_text(size = 18),
    axis.title.y.right = element_text(color = "#3c5488"),
    axis.text.y.right = element_text(color = "#3c5488"),
    axis.title.y.left = element_text(color = "#00a087"),
    axis.text.y.left = element_text(color = "#00a087")
  ) 

thr_plot

# mRNA and protein dynamics for TR are very similar - eg mRNA immediately translated 

```

### Saving plot

```{r}

# ===== CairoPDF =====

# Define dimensions
width_in <- 40 
aspect_ratio <- 3/4 
height_in <- width_in * aspect_ratio

# Convert mm to inche
height_mm <- 4* height_in / 25.4   # ≈ 1.378 inches
width_mm <- 4*width_in / 25.4  # ≈ 1.929 inches
thr_plot

CairoPDF(normalizePath(paste(file_path,'fig1_rna_protein_dynamics_0.pdf', sep = '')),
         width = width_mm, 
         height = height_mm)
print(thr_plot)
dev.off()

# ===== cowplot =====

cowplot::save_plot(plot = thr_plot,
                   base_height = 3.71,
                   base_asp = 1.618,
                   file = normalizePath(paste(file_path, 'fig1_rna_protein_dynamics_1.pdf', sep = '')))

cowplot::save_plot(plot = thr_plot,
                   base_height = 1.1*3.71,
                   base_width = 1.1*3.71*1.618,
                   file = normalizePath(paste(file_path, 'fig1_rna_protein_dynamics-2.pdf', sep = '')))

```
